dali/pipeline/operators/reader/coco_reader_op_test.cc

```diff
@@ -0,0 +1,302 @@
+    23, 54, 39, 62, 17, 6, 73, 58, 19, 41, 71, 61, 79, 32, 12, 27,
+    33, 1, 19, 76, 15, 18, 34, 17, 76, 39, 44, 51, 7, 44, 36, 10, 3,
+    32, 39, 46, 78, 38, 55, 68, 71, 29, 20, 80, 4, 66, 44, 1, 36, 54,
+    24, 23, 20, 51, 46, 11, 71, 46, 49, 16, 12, 24, 54, 33, 46, 16,
+    4, 64, 76, 2, 77, 17, 56, 5, 48, 1, 30, 19, 14, 5, 62, 31,
+  };
+};
+
+typedef ::testing::Types<CPUBackend> TestTypes;
+
+TYPED_TEST_SUITE(CocoReaderTest, TestTypes);
+
+TYPED_TEST(CocoReaderTest, MutuallyExclusiveOptions) {
+  Pipeline pipe(1, 1, 0);
+
+  pipe.AddOperator(
+    this->CocoReaderOpSpec()
+    .AddArg("shuffle_after_epoch", true)
+    .AddArg("stick_to_shard", true));
+
+  EXPECT_THROW(pipe.Build(this->Outputs()), std::runtime_error);
+}
+
+TYPED_TEST(CocoReaderTest, SkipEmpty) {
+  Pipeline pipe(this->SmallCocoSize() - this->EmptyImages(), 1, 0);
+
+  pipe.AddOperator(
+    this->CocoReaderOpSpec().AddArg("skip_empty", true),
+    "coco_reader");
+
+  pipe.Build(this->Outputs());
+
+  ASSERT_EQ(
+    pipe.EpochSize()["coco_reader"],
+    this->SmallCocoSize() - this->EmptyImages());
+
+  DeviceWorkspace ws;
+  pipe.RunCPU();
+  pipe.RunGPU();
+  pipe.Outputs(&ws);
+
+  auto ids = this->CopyIds(ws);
+
+  for (int id = 0; id < this->SmallCocoSize() - this->EmptyImages(); ++id) {
+    ASSERT_EQ(ids[id], id);
+  }
+}
+
+TYPED_TEST(CocoReaderTest, IncludeEmpty) {
+  Pipeline pipe(this->SmallCocoSize(), 1, 0);
+
+  pipe.AddOperator(
+    this->CocoReaderOpSpec(),
+    "coco_reader");
+
+  pipe.Build(this->Outputs());
+
+  ASSERT_EQ(pipe.EpochSize()["coco_reader"], this->SmallCocoSize());
+
+  DeviceWorkspace ws;
+  pipe.RunCPU();
+  pipe.RunGPU();
+  pipe.Outputs(&ws);
+
+  auto ids = this->CopyIds(ws);
+
+  for (int id = 0; id < this->SmallCocoSize(); ++id) {
+    ASSERT_EQ(ids[id], id);
+  }
+
+  this->CheckInstances(ws);
+}
+
+TYPED_TEST(CocoReaderTest, BigSizeThreshold) {
+  Pipeline pipe(this->ImagesWithBigObjects(), 1, 0);
+
+  pipe.AddOperator(
+    this->CocoReaderOpSpec()
+    .AddArg("skip_empty", true)
+    .AddArg("size_threshold", 300.f),
+    "coco_reader");
+
+  pipe.Build(this->Outputs());
+
+  ASSERT_EQ(pipe.EpochSize()["coco_reader"], this->ImagesWithBigObjects());
+
+  DeviceWorkspace ws;
+  pipe.RunCPU();
+  pipe.RunGPU();
+  pipe.Outputs(&ws);
+
+  auto ids = this->CopyIds(ws);
+
+  ASSERT_EQ(ids[0], 3);
+  ASSERT_EQ(ids[1], 17);
+}
+
+TYPED_TEST(CocoReaderTest, ShuffleAfterEpoch) {
+  Pipeline pipe(this->SmallCocoSize(), 1, 0);
+
+  pipe.AddOperator(
+    this->CocoReaderOpSpec()
+    .AddArg("shuffle_after_epoch", true),
+    "coco_reader");
+
+  pipe.Build(this->Outputs());
+
+  DeviceWorkspace ws;
+  pipe.RunCPU();
+  pipe.RunGPU();
+  pipe.Outputs(&ws);
+
+  auto ids_epoch_1 = this->CopyIds(ws);
+
+  pipe.RunCPU();
+  pipe.RunGPU();
+  pipe.Outputs(&ws);
+
+  auto ids_epoch_2 = this->CopyIds(ws);
+
+  for (int id = 0; id < this->SmallCocoSize(); ++id) {
+    if (ids_epoch_1[id] != ids_epoch_2[id])
```

Comment:
`ASSERT_EQ(ids_epoch_1[id], ids_epoch_2[id]);`
and remove the `FAIL()` down below

Reply:
Actually, this test is supposed to check, that `ids_epoch_1` and `ids_epoch_2` are not equal on at least one element (they should be the same array shuffled in different ways). So neither `ASSERT_EQ` nor `ASSERT_FALSE(a == b)`  don't really fit here.
So what I did there was something like: go through both vectors and compare elements. When ypu will find difference thetn `return`, as this is a success. When you reached the end of both vectors and there were no differences, this is `FAIL`

Reply:
can you set a bool and then assert that the bool is your expected value? I don't really like that FAIL call

Reply:
Done